File: Vengeance_BasicModel/ui.py
import chess
import argparse
from movegeneration import next_move

# START FROM SCRATCH 
#DAY1- start and render
#DAY2- get move, get depth
#DAY3- p-progress in state; new maximum available is 0
#DAY4- keyboard interrupts
#DAY5- bam! done finished by deadline

def start():
    board = chess.Board()
    user_side = (
        chess.WHITE if input("Start as [w]hite or [b]lack:\n").lower() == "w" else chess.BLACK
    )

    if user_side == chess.WHITE:
        print(render(board))
        board.push(get_move(board))

    while not board.is_game_over():
        board.push(next_move(get_depth(), board, debug=False))
        print(render(board))
        board.push(get_move(board))

    print(f"\nResult: [w] {board.result()} [b]")

def render(board: chess.Board) -> str:
    """
    print a side-relative chess board using ASCII characters.
    """
    # mapping of chess pieces to ASCII symbols
    ascii_pieces = {
        "R": "R", "N": "N", "B": "B", "Q": "Q", "K": "K", "P": "P",
        "r": "r", "n": "n", "b": "b", "q": "q", "k": "k", "p": "p",
        ".": "."
    }
    
    # Generate the board string
    board_string = str(board).split('\n')

    # Translate the board string into ASCII symbols
    translated_board = [
        ''.join(ascii_pieces.get(char, char) for char in line)
        for line in board_string
    ]

    # Generate the rank labels
    ranks = ["8", "7", "6", "5", "4", "3", "2", "1"]
    
    # Reverse ranks if the turn is black
    if board.turn == chess.BLACK:
        translated_board.reverse()
        ranks.reverse()

    # Construct the display output
    display = [f"  {rank} {line}" for rank, line in zip(ranks, translated_board)]
    display.append("    a b c d e f g h")

    return "\n".join(display)

def get_move(board: chess.Board) -> chess.Move:
    """
    Legality
    """
    move = input(f"\nYour move (e.g. {list(board.legal_moves)[0]}):\n")

    for legal_move in board.legal_moves:
        if move == str(legal_move):
            return legal_move
    return get_move(board)

def get_depth() -> int: # Ok i used chatgpt for syntax
    parser = argparse.ArgumentParser()
    parser.add_argument("--depth", default=3, help="provide an integer (default: 3)")
    args = parser.parse_args()
    return max([1, int(args.depth)])

if __name__ == "__main__":
    try:
        start()
    except KeyboardInterrupt: # Ok i used chatgpt for syntax
        pass

File: Vengeance_AdvancedModel/ui.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stdexcept>
#include <chess/board.hpp> // Example placeholder for a chess library
#include <chess/move.hpp>
#include <chess/engine.hpp> // Example placeholder for chess engine logic

// Function prototypes
void start();
std::string render(const chess::Board& board);
chess::Move getMove(const chess::Board& board);
int getDepth();

void start() {
    chess::Board board;
    char side;
    std::cout << "Start as [w]hite or [b]lack:\n";
    std::cin >> side;
    bool userSide = (side == 'w') ? chess::WHITE : chess::BLACK;

    if (userSide == chess::WHITE) {
        std::cout << render(board) << std::endl;
        board.push(getMove(board));
    }

    while (!board.isGameOver()) {
        board.push(chess::nextMove(getDepth(), board, false));
        std::cout << render(board) << std::endl;
        board.push(getMove(board));
    }

    std::cout << "\nResult: [w] " << board.result() << " [b]" << std::endl;
}

std::string render(const chess::Board& board) {
    // ASCII piece mapping
    std::unordered_map<char, char> asciiPieces = {
        {'R', 'R'}, {'N', 'N'}, {'B', 'B'}, {'Q', 'Q'}, {'K', 'K'}, {'P', 'P'},
        {'r', 'r'}, {'n', 'n'}, {'b', 'b'}, {'q', 'q'}, {'k', 'k'}, {'p', 'p'},
        {'.', '.'}
    };

    // Generate the board string
    std::vector<std::string> boardString = board.toString(); // Assuming a method to get board as string lines

    std::vector<std::string> translatedBoard;
    for (const auto& line : boardString) {
        std::string translatedLine;
        for (char ch : line) {
            translatedLine += asciiPieces[ch];
        }
        translatedBoard.push_back(translatedLine);
    }

    // Rank labels
    std::vector<std::string> ranks = {"8", "7", "6", "5", "4", "3", "2", "1"};
    
    // Reverse ranks if the turn is black
    if (board.getTurn() == chess::BLACK) {
        std::reverse(translatedBoard.begin(), translatedBoard.end());
        std::reverse(ranks.begin(), ranks.end());
    }

    // Construct display output
    std::string display;
    for (size_t i = 0; i < ranks.size(); ++i) {
        display += "  " + ranks[i] + " " + translatedBoard[i] + "\n";
    }
    display += "    a b c d e f g h\n";

    return display;
}

chess::Move getMove(const chess::Board& board) {
    std::string move;
    std::cout << "\nYour move (e.g. " << board.getLegalMoves().front() << "):\n";
    std::cin >> move;

    for (const auto& legalMove : board.getLegalMoves()) {
        if (move == legalMove.toString()) {
            return legalMove;
        }
    }

    return getMove(board); // Retry if move is illegal
}

int getDepth() {
    int depth = 3; // Default value
    if (auto depthStr = std::getenv("DEPTH")) {
        try {
            depth = std::stoi(depthStr);
        } catch (const std::invalid_argument& e) {
            std::cerr << "Invalid depth argument. Using default value.\n";
        }
    }
    return std::max(1, depth);
}

int main() {
    try {
        start();
    } catch (const std::exception& e) {
        std::cerr << "An error occurred: " << e.what() << std::endl;
    }
    return 0;
}
